File: autocontext/src/autocontext/server/trace_gate_review_api.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from autocontext.analytics.cross_runtime_trace_findings import CrossRuntimeTraceFindingReport
from autocontext.analytics.trace_gate_operator_view import build_trace_gate_operator_view
from autocontext.storage.run_paths import resolve_run_root


def build_run_trace_gate_review(*, runs_root: Path, run_id: str) -> dict[str, Any]:
    """Read the operator-facing trace-finding/proposal/gate view for a run."""
    run_root = resolve_run_root(runs_root, run_id)
    return build_trace_gate_operator_view(
        run_id=run_id.strip(),
        report=_load_latest_trace_finding_report(run_root),
        proposals=_load_harness_change_proposals(run_root),
    )


def _load_latest_trace_finding_report(run_root: Path) -> CrossRuntimeTraceFindingReport | None:
    candidates = [
        *_json_files(run_root / "trace-findings"),
        *_json_files(run_root / "trace_findings"),
        run_root / "trace-finding-report.json",
        run_root / "trace_finding_report.json",
    ]
    existing = [path for path in candidates if path.exists()]
    if not existing:
        return None
    existing.sort(key=lambda path: path.stat().st_mtime, reverse=True)
    payload = json.loads(existing[0].read_text(encoding="utf-8"))
    return CrossRuntimeTraceFindingReport.model_validate(payload)


def _load_harness_change_proposals(run_root: Path) -> list[dict[str, Any]]:
    proposals: list[dict[str, Any]] = []
    for path in [*_json_files(run_root / "harness-proposals"), *_json_files(run_root / "harness_proposals")]:
        payload = json.loads(path.read_text(encoding="utf-8"))
        if isinstance(payload, dict):
            proposals.append(payload)
    return proposals


def _json_files(directory: Path) -> list[Path]:
    if not directory.exists():
        return []
    return sorted(path for path in directory.glob("*.json") if path.is_file())

File: ts/src/server/trace-gate-review-api.ts
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { join, resolve } from "node:path";

import {
  buildTraceGateOperatorView,
  type TraceGateOperatorView,
} from "../analytics/trace-gate-operator-view.js";
import { TraceFindingReportSchema, type TraceFindingReport } from "../analytics/trace-findings.js";
import type { HarnessChangeProposal } from "../control-plane/contract/types.js";
import { validateHarnessChangeProposal } from "../control-plane/contract/validators.js";
import type { CockpitApiResponse } from "./cockpit-api.js";

export interface TraceGateReviewApiRoutes {
  getByRunId(runId: string): CockpitApiResponse;
}

export interface BuildTraceGateReviewApiRoutesOptions {
  runsRoot: string;
  loadReport?: (runId: string) => TraceFindingReport | null;
  loadProposals?: (runId: string) => readonly HarnessChangeProposal[];
}

export function buildTraceGateReviewApiRoutes(
  opts: BuildTraceGateReviewApiRoutesOptions,
): TraceGateReviewApiRoutes {
  return {
    getByRunId(runId) {
      let cleanRunId: string;
      try {
        cleanRunId = validateRunId(opts.runsRoot, runId);
      } catch (error) {
        return { status: 422, body: { detail: errorMessage(error) } };
      }

      try {
        const view = buildTraceGateOperatorView({
          run_id: cleanRunId,
          report: opts.loadReport
            ? opts.loadReport(cleanRunId)
            : loadLatestTraceFindingReport(opts.runsRoot, cleanRunId),
          proposals: opts.loadProposals
            ? opts.loadProposals(cleanRunId)
            : loadHarnessChangeProposals(opts.runsRoot, cleanRunId),
        });
        return { status: 200, body: view };
      } catch (error) {
        return { status: 500, body: { detail: errorMessage(error) } };
      }
    },
  };
}

export function loadLatestTraceFindingReport(
  runsRoot: string,
  runId: string,
): TraceFindingReport | null {
  const runRoot = resolveRunRoot(runsRoot, runId);
  const candidates = [
    ...reportFilesInDir(join(runRoot, "trace-findings")),
    ...reportFilesInDir(join(runRoot, "trace_findings")),
    join(runRoot, "trace-finding-report.json"),
    join(runRoot, "trace_finding_report.json"),
  ]
    .filter((path) => existsSync(path))
    .sort((left, right) => statSync(right).mtimeMs - statSync(left).mtimeMs);
  const first = candidates[0];
  return first ? TraceFindingReportSchema.parse(readJson(first)) : null;
}

export function loadHarnessChangeProposals(
  runsRoot: string,
  runId: string,
): HarnessChangeProposal[] {
  const runRoot = resolveRunRoot(runsRoot, runId);
  return [join(runRoot, "harness-proposals"), join(runRoot, "harness_proposals")]
    .flatMap(jsonFilesInDir)
    .map((path) => readHarnessChangeProposal(path));
}

function validateRunId(runsRoot: string, runId: string): string {
  const cleanRunId = runId.trim();
  resolveRunRoot(runsRoot, cleanRunId);
  return cleanRunId;
}

function resolveRunRoot(runsRoot: string, runId: string): string {
  const cleanRunId = runId.trim();
  if (!cleanRunId) {
    throw new Error("run_id is required");
  }
  const root = resolve(runsRoot);
  const candidate = resolve(root, cleanRunId);
  if (candidate === root || !candidate.startsWith(`${root}/`)) {
    throw new Error(`run_id escapes runs root: '${runId}'`);
  }
  return candidate;
}

function readHarnessChangeProposal(path: string): HarnessChangeProposal {
  const payload = readJson(path);
  const validation = validateHarnessChangeProposal(payload);
  if (!validation.valid) {
    throw new Error(`Invalid HarnessChangeProposal at ${path}: ${validation.errors.join("; ")}`);
  }
  return payload as HarnessChangeProposal;
}

function reportFilesInDir(dir: string): string[] {
  return jsonFilesInDir(dir).filter((path) => !path.endsWith(".tmp"));
}

function jsonFilesInDir(dir: string): string[] {
  if (!existsSync(dir)) {
    return [];
  }
  return readdirSync(dir)
    .filter((entry) => entry.endsWith(".json"))
    .map((entry) => join(dir, entry));
}

function readJson(path: string): unknown {
  return JSON.parse(readFileSync(path, "utf-8"));
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

export type { TraceGateOperatorView };
